### PR TITLE
refactor(transaction): add common getters to transaction

### PR DESCRIPTION
**Pull request by meship-starkware**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)

### CODE REVIEW

**Review by meship-starkware — COMMENTED** [review]
@tdelabro I am not sure this is what you had in mind 
please LMK if you have any comments

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229#-:-O37VrAqDskD-i387apt:b3wyo8u)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @noaov1)

File: crates/blockifier/src/transaction/account_transaction.rs [hunk 1]
```diff
@@ -191,6 +197,14 @@ impl AccountTransaction {
         Ok(())
     }
 
+    pub fn get_transaction_hash(&self) -> &TransactionHash {
```
Comment (tdelabro) [inline crates/blockifier/src/transaction/account_transaction.rs:200]: Don't name getters methods using the `get_` prefix

https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

File: crates/blockifier/src/transaction/transaction_execution.rs [hunk 2]
```diff
@@ -97,6 +97,13 @@ impl Transaction {
             _ => unimplemented!(),
         }
     }
+
+    pub fn get_transaction_hash(tx: &Transaction) -> &TransactionHash {
```
Comment (tdelabro) [inline crates/blockifier/src/transaction/transaction_execution.rs:101]: same

**Review by tdelabro — CHANGES_REQUESTED** [review]
Yup this is what I ment.
There may be other ones you could add, have you checked for every fields?

**Review by meship-starkware — COMMENTED** [review]
In the case of the blockier transaction, the only other common field they have is the tx, but everyone is of a different type
So I cannot write a getter for it. In the case of the Transaction object in the starknet API they do have more common fields but they are divided by versions and already have some common getters for some of the fields.  

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229#-:-O4nxPG965blso2oy1Ss:bjawr2a)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 200 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/229#gh-1722517772:-O4o0OV591uHjOoSeEwJ:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/2f07d14a6955dac62bc5a72b35486b424aeaa864/crates/blockifier/src/transaction/account_transaction.rs#L200)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

Don't name getters methods using the `get_` prefix

https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter
</blockquote></details>

Done.
___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 101 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/229#gh-1722517997:-O4o0OAQ9ls39qpiMF0q:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/2f07d14a6955dac62bc5a72b35486b424aeaa864/crates/blockifier/src/transaction/transaction_execution.rs#L101)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.

**Review by meship-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @ArniStarkware, @noaov1, and @tdelabro)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 200 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/229#gh-1722517772:-O4oP6I-5HqvTCdHa1pg:b-8v205a) ([raw file](https://github.com/starkware-libs/sequencer/blob/2f07d14a6955dac62bc5a72b35486b424aeaa864/crates/blockifier/src/transaction/account_transaction.rs#L200)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Someone already added this getter to the account transaction, so I only needed to add it to transaction_execution. 
Also, @ArniStarkware returns TransactionHash inside of  &TransactionHash, so I decided to go with it 
LMK if it might cause a problem. 
https://github.com/starkware-libs/sequencer/blob/7082e88ca91fafa1f202c3592163e1e0b3a8a88b/crates/blockifier/src/transaction/account_transaction.rs#L648

**Comment by ArniStarkware** [comment]
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O4oefa32zQMsp6wViMf):*




_Suggestion:_
```
- abb33f7: refactor(blockifier): add getter for tx_hash to transaction
```

**Review by ArniStarkware — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229#-:-O4oeaGrD5gTbyA0BmMD:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @meship-starkware, @noaov1, and @tdelabro)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O4of4Hb7OVYkNif04Nh):*
Make sure the merged commit message has no typo.



_Code quote:_
```
tramsaction
```

**Review by meship-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 200 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/229#gh-1722517772:-O4oe7n83-G-E6tvBjN0:b-cdel6s) ([raw file](https://github.com/starkware-libs/sequencer/blob/2f07d14a6955dac62bc5a72b35486b424aeaa864/crates/blockifier/src/transaction/account_transaction.rs#L200)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

*instead of?
</blockquote></details>

In the end, I changed my getter to return the object instead of a reference. So the question was for @tdelabro if it fits his need

**Review by ArniStarkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

**Review by ArniStarkware — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

**Review by tdelabro — COMMENTED** [review]
> already have some common getters for some of the fields

Yeah that was what I was referring to. The issue is in the "some".
Those getters have been added erratically, when someone in the Starkware team needed them, without a concern for exhaustivity.
This PR could solve this by checking that every common field is covered by a getter

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229#-:-O5lon2Q9x3QS2giZu7C:b6r48lz)* status: all files reviewed, 2 unresolved discussions (waiting on @meship-starkware)

**Review by meship-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: 0 of 4 files reviewed, 5 unresolved discussions (waiting on @ArniStarkware, @noaov1, and @tdelabro)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 145 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qgnF8CCzt7EPbPwqe:-O5qgnF8CCzt7EPbPwqf:begx3v3) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/blockifier/src/transaction/account_transaction.rs#L145)):*
> ```XML
>         (max_fee, Option<Fee>)
>     );
> ```

I'm not sure about the option here. It might be better to return an error or panic, as I did with the only V3 case.
Just to be clear, I think that all getters who do not have all options need to act the same. When we decide on a better solution, I will refactor all of them to act the same.  
___
*[`crates/starknet_api/src/transaction.rs` line 299 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qiU8dDvYbOW4i8N7z:-O5qiU8dDvYbOW4i8N8-:b-na74ve) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/starknet_api/src/transaction.rs#L299)):*
> ```XML
>         }
>     }
> ```

I am not sure this getter is needed, but all the other fields had one, so I added this.  
___
*[`crates/starknet_api/src/transaction.rs` line 411 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qh__jDJPHHcSzgioT:-O5qh__jDJPHHcSzgioU:bxrgq32) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/starknet_api/src/transaction.rs#L411)):*
> ```XML
>         $(pub fn $field(&self) -> $field_type {
>             match self {
>                 Self::V3(tx) => tx.$field.clone(),
> ```

I am not sure the panic here is a good thing. It does make the code cleaner (no need to wrap in result or option), but I think it better to let the user decide how to handle errors this way and not panic if he accidentally asks a field that does not exist for the wanted version. WDYT?

**Review by ArniStarkware — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: 1 of 4 files reviewed, 7 unresolved discussions (waiting on @meship-starkware, @noaov1, and @tdelabro)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 145 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qgnF8CCzt7EPbPwqe:-O5qtARe1E3cTIGQY_pj:b4xp9l2) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/blockifier/src/transaction/account_transaction.rs#L145)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

I'm not sure about the option here. It might be better to return an error or panic, as I did with the only V3 case.
Just to be clear, I think that all getters who do not have all options need to act the same. When we decide on a better solution, I will refactor all of them to act the same.
</blockquote></details>

Maybe return default class hash.
___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 160 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qrArsDz9WCQ2ZrU6X:-O5qrArsDz9WCQ2ZrU6Y:b2xsm2d) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/blockifier/src/transaction/account_transaction.rs#L160)):*
> ```XML
>             Self::Invoke(tx) => Some(tx.tx.sender_address()),
>         }
>     }
> ```

Deploy account has the sender address as a member of the transaction on an external layer.

Also - make sure we don't have both getters for `contract address` and `sender address`.



_Suggestion:_
```XML
    pub fn sender_address(&self) -> ContractAddress {
        match self {
            Self::Declare(tx) => tx.tx.sender_address(),
            Self::DeployAccount(tx) => tx.contract_address(),
            Self::Invoke(tx) => tx.tx.sender_address(),
        }
    }
```
___
*[`crates/starknet_api/src/executable_transaction.rs` line 160 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qqH8FB2mxl1kzVF1Y:-O5qqH8FB2mxl1kzVF1Z:b3blegx) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/starknet_api/src/executable_transaction.rs#L160)):*
> ```XML
>         (fee_data_availability_mode, DataAvailabilityMode),
>         (paymaster_data, PaymasterData),
>         (max_fee, Option<Fee>)
> ```

I don't like the fact that `executable transaction` can return max fee.

the executable transaction should be a creature of v3 transaction only (I know it is not, this is the ideal).



_Code quote:_
```XML
        (max_fee, Option<Fee>)
```
___
*[`crates/starknet_api/src/transaction.rs` line 299 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qiU8dDvYbOW4i8N7z:-O5qscXJ-67p4wsvoiPz:be2yn2d) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/starknet_api/src/transaction.rs#L299)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

I am not sure this getter is needed, but all the other fields had one, so I added this.
</blockquote></details>

I sent you a DM about it.
___
*[`crates/starknet_api/src/transaction.rs` line 411 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qh__jDJPHHcSzgioT:-O5qsvDD7CGyYlWFub_4:bmj3pm7) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/starknet_api/src/transaction.rs#L411)):*
<details><summary><i>Previously, meship-starkware (Meshi Peled) wrote…</i></summary><blockquote>

I am not sure the panic here is a good thing. It does make the code cleaner (no need to wrap in result or option), but I think it better to let the user decide how to handle errors this way and not panic if he accidentally asks a field that does not exist for the wanted version. WDYT?
</blockquote></details>

Maybe you can return the default, but I don't like that either...

**Review by meship-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: 1 of 4 files reviewed, 7 unresolved discussions (waiting on @ArniStarkware, @noaov1, and @tdelabro)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 160 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qrArsDz9WCQ2ZrU6X:-O5s1G2b4OCDvsfcchlH:b-8a2cdq) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/blockifier/src/transaction/account_transaction.rs#L160)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Deploy account has the sender address as a member of the transaction on an external layer.

Also - make sure we don't have both getters for `contract address` and `sender address`.
</blockquote></details>

Done.
in this pr https://reviewable.io/reviews/starkware-libs/sequencer/699

**Review by meship-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: 1 of 4 files reviewed, 5 unresolved discussions (waiting on @ArniStarkware, @noaov1, and @tdelabro)

___
*[`crates/blockifier/src/transaction/account_transaction.rs` line 145 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qgnF8CCzt7EPbPwqe:-O5s1Nqm1RIz-YJPpU1d:bhxsi8m) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/blockifier/src/transaction/account_transaction.rs#L145)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

Maybe return default class hash.
</blockquote></details>

added relevant comment in this PR https://reviewable.io/reviews/starkware-libs/sequencer/699
___
*[`crates/starknet_api/src/executable_transaction.rs` line 160 at r4](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O5qqH8FB2mxl1kzVF1Y:-O5s4Ief-qqqkgw6kt7J:b-7xyddj) ([raw file](https://github.com/starkware-libs/sequencer/blob/bb6dd7f915a4468d6e980e25c95e2c4412e0051d/crates/starknet_api/src/executable_transaction.rs#L160)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

I don't like the fact that `executable transaction` can return max fee.

the executable transaction should be a creature of v3 transaction only (I know it is not, this is the ideal).
</blockquote></details>

will do in this PR https://reviewable.io/reviews/starkware-libs/sequencer/699

**Review by ArniStarkware — APPROVED** [review]
:lgtm: - post expansion of the scope of this PR, and the split.

> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229#-:-O5sC1nB4FTeMh7iKaxB:badegkg)* status: all files reviewed, 3 unresolved discussions (waiting on @meship-starkware, @noaov1, and @tdelabro)

**Review by ArniStarkware — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

**Review by ArniStarkware — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

**Comment by ArniStarkware** [comment]
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 52 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O6Gorjc9eBv6aqxAOEU:-O6Gorjc9eBv6aqxAOEV:b29cq9o) ([raw file](https://github.com/starkware-libs/sequencer/blob/ad686eeb1c6ed26991bc9135138236f464ba690a/crates/blockifier/src/transaction/transaction_execution.rs#L52)):*
> ```XML
>         }
>     }
>     pub fn from_api(
> ```

spaces.



_Suggestion:_
```XML
    pub fn nonce(&self) -> Nonce {
        match self {
            Self::AccountTransaction(tx) => tx.nonce(),
            Self::L1HandlerTransaction(tx) => tx.tx.nonce,
        }
    }
    
    pub fn sender_address(&self) -> ContractAddress {
        match self {
            Self::AccountTransaction(tx) => tx.sender_address(),
            Self::L1HandlerTransaction(tx) => tx.tx.contract_address,
        }
    }
    
    pub fn from_api(
```

**Review by ArniStarkware — COMMENTED** [review]
> Reviewed 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 3 unresolved discussions (waiting on @meship-starkware, @noaov1, and @tdelabro)

**Review by meship-starkware — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @ArniStarkware, @noaov1, and @tdelabro)

___
*[`crates/blockifier/src/transaction/transaction_execution.rs` line 52 at r9](https://reviewable.io/reviews/starkware-libs/sequencer/229#-O6Gorjc9eBv6aqxAOEU:-O6Gw53OEJjT8dnJuSI3:b-896fix) ([raw file](https://github.com/starkware-libs/sequencer/blob/ad686eeb1c6ed26991bc9135138236f464ba690a/crates/blockifier/src/transaction/transaction_execution.rs#L52)):*
<details><summary><i>Previously, ArniStarkware (Arnon Hod) wrote…</i></summary><blockquote>

spaces.
</blockquote></details>

Done.

**Review by ArniStarkware — APPROVED** [review]
> Reviewed 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

**Review by ArniStarkware — APPROVED** [review]
> Reviewed 2 of 3 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1 and @tdelabro)

**Review by tdelabro — COMMENTED** [review]
> Reviewed 2 of 3 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: all files reviewed, 2 unresolved discussions (waiting on @noaov1)

**Review by meship-starkware — COMMENTED** [review]
> Dismissed @tdelabro from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @noaov1)

**Review by noaov1 — APPROVED** [review]
> Reviewed 2 of 3 files at r5, 1 of 3 files at r11, 1 of 1 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/229)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @meship-starkware)

**Pull request merged**